pkg/k8s/ipam_state.go

```diff
@@ -0,0 +1,158 @@
+	}
+
+	namespace, podName := items[0], items[1]
+	i := strings.LastIndex(podName, "-")
+	if i < 1 {
+		log.Infof("%s is not sts pod: contains more than 1 dashes", fullPodName)
+		return false, nil
+	}
+
+	stsName := podName[:i]
+	if _, err := strconv.Atoi(podName[i+1:]); err != nil {
+		log.Infof("%s is not sts pod: index not found", fullPodName)
+		return false, nil
+	}
+
+	// call K8S API
+	_, err := Client().AppsV1().StatefulSets(namespace).Get(stsName,
+		metav1.GetOptions{})
+	if err != nil {
+		switch err.Error() {
+		case `statefulsets.apps "` + stsName + `" not found`:
+			log.Infof("%s is not sts pod: sts not found in k8s", fullPodName)
+			return false, nil
+		default:
+			return false, err
+		}
+	}
+
+	log.Infof("pod %s belongs to a sts", fullPodName)
+	return true, nil
+}
+
+// IsStsPodDeleted determines if the pod is deleted from the node by retrieving
+// the metadata in k8s API
+// pod name format: <namespace>/<stsName>-<podIndex>
+func IsStsPodDeleted(nodeName string, fullPodName string) (bool, error) {
+	// extract sts name
+	items := strings.Split(fullPodName, "/")
+	if len(items) != 2 {
+		return false, fmt.Errorf("unexpected fullPodName %s", fullPodName)
+	}
+
+	ns, podName := items[0], items[1]
+	i := strings.LastIndex(podName, "-")
+	if i < 1 {
+		return false, fmt.Errorf("mal-formed pod name %s", podName)
+	}
+
+	stsName := podName[:i]
+	podIndex, err := strconv.Atoi(podName[i+1:])
+	if err != nil {
+		return false, err
+	}
+
+	// get info through K8S API
+	log.Infof("check if %s/%s still exists in apiserver", ns, stsName)
+	sts, err := Client().AppsV1().StatefulSets(ns).Get(stsName,
+		metav1.GetOptions{})
+	if err != nil {
+		switch err.Error() {
+		case `statefulsets.apps "` + stsName + `" not found`:
+			log.Infof("Statefulset %s has been deleted from K8S", fullPodName)
+			return true, nil
+		default:
+			return false, err
+		}
+	}
+
+	const (
+		foreground = "foregroundDeletion"
+	)
+
+	// If the resource is being deleted with PropagationPolicy foreground,
+	// the replicas field will remain unchanged, so we need to handle such cases here.
+	finalizers := (*sts).ObjectMeta.Finalizers
+	if len(finalizers) > 0 {
+		log.Infof("PropagationPolicy case")
+		propagationPolicy := finalizers[0]
+		if propagationPolicy == foreground {
+			log.Infof("PropagationPolicy == foreground case")
+			return true, nil
+		}
+	}
+
+	replicas := int(*sts.Spec.Replicas)
+	log.Infof("sts replicas: %d, pod index: %d", replicas, podIndex)
+	if replicas > 0 && podIndex < replicas {
+		return false, nil // valid index, pod still in sts replicas
+	}
+
+	return true, nil // pod truly deleted from sts replicas
+}
+
+// GetNodeStsPods calls K8S API and returns all sts pods on this node
+func GetNodeStsPods(nodeName string) ([]string, error) {
+	podNames := []string{}
+	options := metav1.ListOptions{FieldSelector: "spec.nodeName=" + nodeName}
+	podList, err := Client().CoreV1().Pods("").List(options)
+	if err != nil {
+		return podNames, err
+	}
```

Comment:
There's a function called `func (k *K8sWatcher) GetCachedPod(namespace, name string) (*types.Pod, error) {` which lives under `pkg/k8s/watchers/pod.go`. Please use it instead of doing API calls directly.

Reply:
I can't find this method in the source code, should i need to rebase to HEAD of master first?

Reply:
Yes, it was recently merged